lib/ansible/modules/cloud/amazon/redshift_security_group_facts.py

```diff
@@ -0,0 +1,229 @@
+    tags:
+      env: production
+  register: redshift_security_groups
+
+# Fail if no groups are found
+- redshift_security_group_facts:
+    name: public-webserver-security_group
+  register: redshift_security_groups
+  failed_when: "{{ security_groups.results | length == 0 }}"
+
+# Fail if more than 1 group is found
+- redshift_security_group_facts:
+    name: public-webserver-security_group
+  register: redshift_security_groups
+  failed_when: "{{ security_groups.results | length > 1 }}"
+'''
+
+RETURN = '''
+---
+cluster_security_group_name:
+    description: Name of the Redshift cluster security group
+    returned: success
+    type: string
+    sample: "default"
+description:
+    description: Description of the Redshift cluster security group
+    returned: success
+    type: string
+    sample: "default"
+ec2_security_groups:
+    description: A list of EC2 security groups that are permitted to access the cluster.
+    returned: success
+    type: list
+    sample: [
+        {
+            "ec2_security_group_name": "launch-wizard-3",
+            "ec2_security_group_owner_id": "1234567890",
+            "status": "authorized",
+            "tags": []
+        },
+        {
+            "ec2_security_group_name": "quicklaunch-1",
+            "ec2_security_group_owner_id": "1234567890",
+            "status": "authorized",
+            "tags": []
+        }
+    ]
+ip_ranges:
+    description: A list of IP ranges (CIDR blocks) that are permitted to access the cluster.
+    returned: success
+    type: list
+    sample: [
+        {
+            "cidrip": "1.1.1.1/32",
+            "status": "authorized",
+            "tags": []
+        },
+        {
+            "cidrip": "2.2.2.2/32",
+            "status": "authorized",
+            "tags": []
+        }
+    ]
+tags:
+    description: List of tags for the ASG, and whether or not each tag propagates to instances at launch.
+    returned: success
+    type: list
+    sample: [
+        {
+            "key": "Env",
+            "value": "prd"
+        },
+        {
+            "key": "Stack",
+            "value": "db"
+        }
+    ]
+'''
+
+import re
+# Import module snippets
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import ec2_argument_spec, boto3_conn, get_aws_connection_info
+from ansible.module_utils.ec2 import HAS_BOTO3, camel_dict_to_snake_dict
+
+try:
+    from botocore.exceptions import ClientError
+except ImportError:
+    pass  # will be picked up from imported HAS_BOTO3
+
+
+def match_csg_tags(tags_to_match, csg):
+    for key, value in tags_to_match.items():
+        for tag in csg['Tags']:
+            if key == tag['Key'] and value == tag['Value']:
+                return True
+
+    return False
+
+
+def find_csgs(conn, module, name=None, tags=None):
+
+    try:
+        csgs_paginator = conn.get_paginator('describe_cluster_security_groups')
+        csgs = csgs_paginator.paginate().build_full_result()
```

Comment:
Instead of paginating over all the security groups, it would be better to restrict what AWS returns using the TagKeys/TagValues parameters on the [describe_cluster_security_groups](http://boto3.readthedocs.io/en/latest/reference/services/redshift.html#Redshift.Client.describe_cluster_security_groups) method.